File: src/praisonai-agents/praisonaiagents/memory/learn/protocols.py
"""
Learn Protocol Definitions.

Provides Protocol interfaces for learning implementations.
Enables custom learning backends and extensibility.
"""
from typing import Any, Dict, List, Optional, Protocol, runtime_checkable

# Re-export LearnMode from feature_configs (canonical location)
# This avoids duplication while maintaining backward compatibility
from ...config.feature_configs import LearnMode


@runtime_checkable
class LearnProtocol(Protocol):
    """
    Protocol for learning store implementations.
    
    Enables custom backends (file, SQLite, Redis, MongoDB, etc.)
    while maintaining a consistent interface.
    
    Example:
        ```python
        class RedisLearnStore:
            def add(self, content: str, metadata=None) -> Dict[str, Any]:
                # Store in Redis
                return {"id": "...", "content": content}
            
            def search(self, query: str, limit: int = 10) -> List[Dict[str, Any]]:
                # Search Redis
                return []
            
            def list_all(self, limit: int = 100) -> List[Dict[str, Any]]:
                return []
            
            def delete(self, entry_id: str) -> bool:
                return True
            
            def clear(self) -> int:
                return 0
        ```
    """
    
    def add(
        self,
        content: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Add a new entry to the store.
        
        Args:
            content: The content to store
            metadata: Optional metadata dictionary
            
        Returns:
            Dictionary with entry details including 'id'
        """
        ...
    
    def search(
        self,
        query: str,
        limit: int = 10,
    ) -> List[Dict[str, Any]]:
        """
        Search entries by query.
        
        Args:
            query: Search query string
            limit: Maximum results to return
            
        Returns:
            List of matching entries
        """
        ...
    
    def list_all(
        self,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """
        List all entries.
        
        Args:
            limit: Maximum entries to return
            
        Returns:
            List of all entries (most recent first)
        """
        ...
    
    def get(
        self,
        entry_id: str,
    ) -> Optional[Dict[str, Any]]:
        """
        Get entry by ID.
        
        Args:
            entry_id: Entry identifier
            
        Returns:
            Entry dict or None if not found
        """
        ...
    
    def update(
        self,
        entry_id: str,
        content: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Optional[Dict[str, Any]]:
        """
        Update an existing entry.
        
        Args:
            entry_id: Entry identifier
            content: New content
            metadata: Optional metadata to merge
            
        Returns:
            Updated entry or None if not found
        """
        ...
    
    def delete(
        self,
        entry_id: str,
    ) -> bool:
        """
        Delete an entry.
        
        Args:
            entry_id: Entry identifier
            
        Returns:
            True if deleted, False if not found
        """
        ...
    
    def clear(self) -> int:
        """
        Clear all entries.
        
        Returns:
            Number of entries cleared
        """
        ...


@runtime_checkable
class AsyncLearnProtocol(Protocol):
    """
    Async Protocol for learning store implementations.
    
    For backends that support async operations.
    """
    
    async def aadd(
        self,
        content: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Async add entry."""
        ...
    
    async def asearch(
        self,
        query: str,
        limit: int = 10,
    ) -> List[Dict[str, Any]]:
        """Async search entries."""
        ...
    
    async def alist_all(
        self,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """Async list all entries."""
        ...
    
    async def aget(
        self,
        entry_id: str,
    ) -> Optional[Dict[str, Any]]:
        """Async get entry by ID."""
        ...
    
    async def aupdate(
        self,
        entry_id: str,
        content: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Optional[Dict[str, Any]]:
        """Async update entry."""
        ...
    
    async def adelete(
        self,
        entry_id: str,
    ) -> bool:
        """Async delete entry."""
        ...
    
    async def aclear(self) -> int:
        """Async clear all entries."""
        ...


@runtime_checkable
class LearnManagerProtocol(Protocol):
    """
    Protocol for LearnManager implementations.
    
    Defines the interface for managing learning across multiple stores.
    """
    
    def capture_persona(self, content: str, category: str = "general") -> Optional[Dict[str, Any]]:
        """Capture user preference or profile information."""
        ...
    
    def capture_insight(self, content: str, source: str = "interaction") -> Optional[Dict[str, Any]]:
        """Capture an observation or learning."""
        ...
    
    def capture_pattern(self, pattern: str, pattern_type: str = "general") -> Optional[Dict[str, Any]]:
        """Capture a reusable knowledge pattern."""
        ...
    
    def get_learning_context(self, limit_per_store: int = 5) -> Dict[str, List[Dict[str, Any]]]:
        """Get learning context from all enabled stores."""
        ...
    
    def search(self, query: str, limit: int = 10) -> Dict[str, List[Dict[str, Any]]]:
        """Search across all enabled stores."""
        ...
    
    def to_system_prompt_context(self) -> str:
        """Generate context for system prompt injection."""
        ...
    
    def process_conversation(
        self,
        messages: List[Dict[str, Any]],
        llm: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Process a conversation to extract learnings automatically.
        
        Args:
            messages: Conversation messages
            llm: Optional LLM model to use for extraction
            
        Returns:
            Dictionary with extracted learnings
        """
        ...


__all__ = [
    'LearnMode',
    'LearnProtocol',
    'AsyncLearnProtocol',
    'LearnManagerProtocol',
]

File: src/praisonai-ts/src/memory/learn/protocols.ts
/**
 * Learn protocol definitions.
 *
 * Python parity: praisonaiagents/memory/learn/protocols.py
 *
 * Provides the interfaces that learning store and manager implementations
 * satisfy, so custom backends (file, SQLite, Redis, MongoDB, ...) plug in
 * behind a consistent surface. `LearnMode` is re-exported from the config
 * module exactly as Python re-exports it from `config.feature_configs`
 * (protocols.py:9-11).
 *
 * Also home to the learn error classes (no direct Python counterpart: Python
 * surfaces a missing backend as `ImportError`, see
 * `storage/backends.py:337-380`). They extend `PraisonAIError` so callers can
 * catch the whole SDK hierarchy uniformly.
 */

import { LearnMode } from '../../config/index';
import { PraisonAIError, type PraisonAIErrorOptions } from '../../errors';

export { LearnMode };

// ============================================================================
// Entry shapes
// ============================================================================

/**
 * Serialised form of a learning entry. Keys are snake_case so JSON files
 * written by the two SDKs are interchangeable.
 * Python parity: praisonaiagents/memory/learn/stores.py:38-47 (`LearnEntry.to_dict`)
 */
export interface LearnEntryData {
  id: string;
  content: string;
  metadata: Record<string, any>;
  created_at: string;
  updated_at: string;
  use_count: number;
  last_used: string | null;
}

/** Anything that can serialise itself to {@link LearnEntryData}. */
export interface LearnEntryConvertible {
  toDict(): LearnEntryData;
}

/**
 * What a store may return: the built-in stores return `LearnEntry` objects
 * (stores.py:139-243), while the protocol docstring shows plain dicts
 * (protocols.py:22-41). The manager normalises both via `toDict()`.
 */
export type LearnEntryLike = LearnEntryData | LearnEntryConvertible;

/** Chat message accepted by `processConversation` (manager.py:509). */
export interface LearnMessage {
  role?: string;
  content?: unknown;
  [key: string]: unknown;
}

// ============================================================================
// Protocols
// ============================================================================

/**
 * Protocol for learning store implementations.
 * Python parity: praisonaiagents/memory/learn/protocols.py:14-149 (`LearnProtocol`)
 */
export interface LearnProtocol {
  /** protocols.py:44-59 */
  add(content: string, metadata?: Record<string, any> | null): LearnEntryLike;
  /** protocols.py:61-76 (`limit` default 10) */
  search(query: string, limit?: number): LearnEntryLike[];
  /** protocols.py:78-91 (`list_all`, `limit` default 100) */
  listAll(limit?: number): LearnEntryLike[];
  /** protocols.py:93-106 */
  get(entryId: string): LearnEntryLike | null;
  /** protocols.py:108-125 */
  update(entryId: string, content: string, metadata?: Record<string, any> | null): LearnEntryLike | null;
  /** protocols.py:127-140 */
  delete(entryId: string): boolean;
  /** protocols.py:142-149 */
  clear(): number;
}

/**
 * Async protocol for learning store implementations.
 * Python parity: praisonaiagents/memory/learn/protocols.py:152-208 (`AsyncLearnProtocol`)
 */
export interface AsyncLearnProtocol {
  /** protocols.py:160-166 */
  aadd(content: string, metadata?: Record<string, any> | null): Promise<LearnEntryLike>;
  /** protocols.py:168-174 */
  asearch(query: string, limit?: number): Promise<LearnEntryLike[]>;
  /** protocols.py:176-181 (`alist_all`) */
  alistAll(limit?: number): Promise<LearnEntryLike[]>;
  /** protocols.py:183-188 */
  aget(entryId: string): Promise<LearnEntryLike | null>;
  /** protocols.py:190-197 */
  aupdate(entryId: string, content: string, metadata?: Record<string, any> | null): Promise<LearnEntryLike | null>;
  /** protocols.py:199-204 */
  adelete(entryId: string): Promise<boolean>;
  /** protocols.py:206-208 */
  aclear(): Promise<number>;
}

/**
 * Result of `processConversation`.
 * Python parity: praisonaiagents/memory/learn/manager.py:513-519 and :597-608
 */
export interface ProcessConversationResult {
  persona: string[];
  insights: string[];
  patterns: string[];
  improvements: string[];
  stored: Record<string, number>;
  error?: string;
}

/**
 * Protocol for LearnManager implementations.
 * Python parity: praisonaiagents/memory/learn/protocols.py:211-258 (`LearnManagerProtocol`)
 *
 * `processConversation` is async here: the TypeScript LLM client has no
 * synchronous call path (Python's `LLM.get_response` is sync).
 */
export interface LearnManagerProtocol {
  /** protocols.py:219-221 */
  capturePersona(content: string, category?: string): LearnEntryLike | null;
  /** protocols.py:223-225 */
  captureInsight(content: string, source?: string): LearnEntryLike | null;
  /** protocols.py:227-229 */
  capturePattern(pattern: string, patternType?: string): LearnEntryLike | null;
  /** protocols.py:231-233 (`get_learning_context`, `limit_per_store` default 5) */
  getLearningContext(limitPerStore?: number): Record<string, LearnEntryData[]>;
  /** protocols.py:235-237 */
  search(query: string, limit?: number): Record<string, LearnEntryData[]>;
  /** protocols.py:239-241 (`to_system_prompt_context`) */
  toSystemPromptContext(): string;
  /** protocols.py:243-258 (`process_conversation`) */
  processConversation(messages: LearnMessage[], llm?: string | null): Promise<ProcessConversationResult>;
}

// ============================================================================
// Errors
// ============================================================================

/**
 * Base class for learn-subsystem failures. Extends {@link PraisonAIError}
 * (errors.ts) so it carries `agentId`, `runId`, `errorCategory` and
 * `context` like every other SDK error.
 */
export class LearnError extends PraisonAIError {
  constructor(message: string, options: PraisonAIErrorOptions = {}) {
    super(message, options);
    Object.setPrototypeOf(this, new.target.prototype);
    this.name = 'LearnError';
  }
}

/**
 * Raised when a learn storage backend cannot be constructed in this runtime:
 * the optional driver is not installed, or the backend has not been ported.
 *
 * Python analogue: the `ImportError` raised by
 * `storage/backends.py:337-380` (`_LazyStorageModule`) when Redis / MongoDB
 * are requested without the `praisonai` package. `LearnManager` treats it
 * exactly as Python does (manager.py:185-213): logs a warning and falls back
 * to the FILE backend.
 */
export class LearnBackendNotAvailableError extends LearnError {
  /** Backend name as passed in `LearnConfig.backend` (e.g. `"redis"`). */
  readonly backend: string;
  /** Human-readable hint on how to make the backend available. */
  readonly installHint: string | null;

  constructor(backend: string, message: string, installHint: string | null = null, options: PraisonAIErrorOptions = {}) {
    super(message, {
      ...options,
      context: { backend, install_hint: installHint, ...(options.context ?? {}) },
    });
    Object.setPrototypeOf(this, new.target.prototype);
    this.name = 'LearnBackendNotAvailableError';
    this.backend = backend;
    this.installHint = installHint;
  }
}

/** Normalise a store result to its dict form (handles both protocol shapes). */
export function toLearnEntryDict(entry: LearnEntryLike): LearnEntryData {
  if (entry && typeof (entry as LearnEntryConvertible).toDict === 'function') {
    return (entry as LearnEntryConvertible).toDict();
  }
  return entry as LearnEntryData;
}
